core/src/main/scala/kafka/server/metadata/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,200 @@
+}
+
+sealed trait BrokerMetadataEvent {}
+
+final case object StartupEvent extends BrokerMetadataEvent
+
+/**
+ * A batch of messages from the metadata log
+ *
+ * @param apiMessages the batch of messages
+ * @param lastMetadataOffset the metadata offset of the last message in the batch
+ */
+final case class MetadataEvent(apiMessages: List[ApiMessage], lastMetadataOffset: Long) extends BrokerMetadataEvent
+
+/**
+ * An event that occurs when the broker heartbeat receives a successful registration response.
+ * It will only occur once in the lifetime of the broker process, and it will occur before
+ * any metadata log message batches appear.  The manager injects this event into the event stream,
+ * and processors will receive it immediately.
+ *
+ * @param brokerEpoch the epoch assigned to the broker by the active controller
+ */
+final case class OutOfBandRegisterLocalBrokerEvent(brokerEpoch: Long) extends BrokerMetadataEvent
+
+/**
+ * An event that occurs when either:
+ *
+ * 1) the local broker's heartbeat is unable to contact the active controller within the
+ * defined lease duration and loses its lease
+ * 2) the local broker is told by the controller that it should be in the fenced state
+ *
+ * The manager injects this event into the event stream such that processors receive it as soon as they finish their
+ * current message batch, if any, otherwise they receive it immediately.
+ *
+ * @param brokerEpoch the broker epoch that was fenced
+ */
+final case class OutOfBandFenceLocalBrokerEvent(brokerEpoch: Long) extends BrokerMetadataEvent
+
+/*
+ * WakeupEvent is necessary for the case when the event thread is blocked in queue.take() and we need to wake it up.
+ * This event has no other semantic meaning and should be ignored when seen.  It is useful in two specific situations:
+ * 1) when an out-of-band message is available; and 2) when shutdown needs to occur.  The presence of this message
+ * ensures these other messages are seen quickly when no other messages are in the queue.
+ */
+final case object WakeupEvent extends BrokerMetadataEvent
+
+class QueuedEvent(val event: BrokerMetadataEvent, val enqueueTimeMs: Long) {
+  override def toString: String = {
+    s"QueuedEvent(event=$event, enqueueTimeMs=$enqueueTimeMs)"
+  }
+}
+
+trait MetadataEventProcessor {
+  def currentMetadataOffset(): Long
+  def processStartup(): Unit
+  def process(apiMessages: List[ApiMessage], lastMetadataOffset: Long): Unit
+  def process(registerLocalBrokerEvent: OutOfBandRegisterLocalBrokerEvent): Unit
+  def process(fenceLocalBrokerEvent: OutOfBandFenceLocalBrokerEvent): Unit
+}
+
+class BrokerMetadataEventManager(config: KafkaConfig,
+                                 time: Time,
+                                 processor: MetadataEventProcessor,
+                                 eventQueueTimeHistogramTimeoutMs: Long = 300000) extends KafkaMetricsGroup {
+
+  private val blockingQueue = new LinkedBlockingQueue[QueuedEvent]
```

Comment:
I think 3 different queues plus a volatile is too much here.

I'd do one of two things:
1. just use KafkaEventQueue.  This allows you to insert events to the beginning of the queue as well as the end, which avoids the need to keep a separate "out of band queue" that gets processed first.  It also manages the thread for you, so you don't need to start and manage your own thread.
2. Use a lock and just keep around the data structures you need rather than messing with concurrents and volatiles.  You won't ever need to hold the lock for long.

Reply:
By the way, I didn't implement a KafkaEventQueue#size function yet, but it would be simple to do, for the purpose of metrics.

Reply:
I changed it to a single `LinkedBlockingDeque` but then a test failed because of the case where 2 out-of-band events come in before the current message batch finishes processing.  What happened is the first one was prepended, then the second one was prepended -- and then they are out-of-order because the second one will be processed before the first one.  So a deque will only work if I check for and deal with this case, which can't be done atomically without a lock.  So one possibility is a single non-concurrent Deque (e.g. LinkedList) with a lock.   The other possibility is a queue with a separate out-of-band queue as I've already done, but I can use a lock so that the data structures aren't concurrent -- though the size will still need to be `@volatile` (in both cases).  Non-concurrent data structures eliminate the need for draining to a buffer queue, of course.

I'll go with a lock and a single dequeue and the special out-of-order checking logic mentioned above fir out-of-band messages since its easy to do and it would be unusual for that case to actually occur.

Reply:
That's a good point.  Out-of-band events need to be handled in order relative to each other.

> I'll go with a lock and a single deque and the special out-of-order checking logic mentioned above for out-of-band messages since its easy to do and it would be unusual for that case to actually occur.

OK.